### DISCUSSION TITLE
signing and sending the transactions

### DISCUSSION THREAD

**Discussion by degenSumi** [aptos-labs/aptos-developer-discussions#423] [Questions]
### Discord user ID

sumibisht

### Describe your question in detail.

just a quick question, now how can i send this transaction

const txn = await aptos.transaction.build.simple({
sender: "0xbe5cd08d8fe346b9b55dba64fc09e50dda92a8bd16743652ec8b156821123310",
data: {
function: "0xc7efb4076dbe143cbcd98cfaaa929ecfc8f299203dfff63b95ccb6bfe19850fa::router::swap_exact_input",
typeArguments: [parseTypeTag(APTOS_COIN), parseTypeTag(DEVNET_BNB)],
functionArguments: [new U64(APTOS_COIN_AMOUNT), new U64(5)],
},
});
console.log(txn);

// Now i want to encode this tx and send it over the internet so that anyone else can sign it

const signedTx = aptos.transaction.sign({
signer:signer,
transaction: txn
});

// Now i want to send this signed tx so that any one else can send this
console.log(signedTx.deserialize(), txn);  // this does'nt work

please let me know as i can't find any functions for it on the sdk

### Which operating system are you using?

macOS

### Which SDK or tool are you using? (if any)

TypeScript SDK

**Comment by degenSumi** [comment]
when i do this:

`const txn = await aptos.transaction.build.simple({
    sender: "0xbe5cd08d8fe346b9b55dba64fc09e50dda92a8bd16743652ec8b156821123310",
    data: {
      function: "0xc7efb4076dbe143cbcd98cfaaa929ecfc8f299203dfff63b95ccb6bfe19850fa::router::swap_exact_input",
      typeArguments: [parseTypeTag(APTOS_COIN), parseTypeTag(DEVNET_BNB)],
      functionArguments: [new U64(APTOS_COIN_AMOUNT), new U64(5)],
    },
  });
  console.log(txn);

  const bytes = txn.rawTransaction.bcsToBytes();`




  console.log(Buffer.from(bytes).toString("base64"));
            console.log(await submitSignedBCSTransaction(Buffer.from(bytes, "base64")));

getting this:

`ApiError: {"message":"Failed to deserialize input into SignedTransaction: unexpected end of input","error_code":"invalid_input","vm_error_code":null}
    at _AptosClient.wrapper (/Users/sumit/Desktop/peer/node_modules/aptos/dist/index.js:4511:15)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Object.signTransactionAptos (/Users/sumit/Desktop/peer/src/adapters/Wallet/signTransaction/Aptos.js:25:25)
    at async exports.signTransactionAptos (/Users/sumit/Desktop/peer/src/adapters/Wallet/signTransaction/index.js:51:21)
    at async Wallet.signTransaction (/Users/sumit/Desktop/peer/src/adapters/Wallet/index.js:50:29)
    at async main (/Users/sumit/Desktop/peer/test/aptosTest.js:17:22) {
  status: 400,
  errorCode: 'invalid_input',
  vmErrorCode: null,
  chainId: '1401'
}`

**Comment by 0xmaayan** [comment]
Hi, here are some examples that might help you with what you are trying to achieve

Shows how one can do `external signing` https://github.com/aptos-labs/aptos-ts-sdk/blob/main/examples/typescript/external_signing.ts#L152

Shows how can `send to a server to submit` (it contains some sponsor transaction logic, but can ignore it) https://github.com/aptos-labs/aptos-ts-sdk/blob/main/examples/typescript-esm/sponsored_transactions/server_signs_and_submit.ts#L4-L6

**Reply by degenSumi** [reply]
Hi, thanks for the reply, 
It's really helpful,
Can you help me if I want to sign the transaction old way using js SDK (aptos) where we get signed serialized transaction string that we can send over the network, as in these I have to send two objects one the transaction second the authentication object

**Comment by degenSumi** [comment]
Hi @0xmaayan 

I am looking for something like this:
  `signTransaction(accountFrom: AptosAccount, rawTransaction: RawTransaction): Promise<Uint8Array>;
    /**
     * Event types are globally identifiable by an account `address` and
     * monotonically increasing `creation_number`, one per event type emitted
     * to the given account. This API returns events corresponding to that
     * that event type.
     * @param address Hex-encoded 32 byte Aptos account, with or without a `0x` prefix,
     * for which events are queried. This refers to the account that events were emitted
     * to, not the account hosting the move module that emits that event type.
     * @param creationNumber Creation number corresponding to the event type.
     * @returns Array of events assotiated with the given account and creation number.
     */`

**Comment by degenSumi** [comment]
Right now i am getting this

`
message:
'Unknown transaction type.'
stack:
'Error: Unknown transaction type.\n `

as i am using type sdk for the raw tx prepare and js sdk for signing <as i can generate a single rawtransaction through it>

**Comment by 0xmaayan** [comment]
Any reason of using both TS and JS SDKs? To generate a transaction and get the signing message you can easily do it with only the TS SDK. 
Keep in mind the JS SDK is not maintained and missing some features and logic and soon will be deprecated.

Check out https://github.com/aptos-labs/aptos-ts-sdk/blob/main/tests/unit/signingMessage.test.ts#L11

Generate a transaction https://github.com/aptos-labs/aptos-ts-sdk/blob/main/src/api/transactionSubmission/build.ts#L31
```
const transaction = await aptos.transaction.build.simple({....})
```

Get the signing message https://github.com/aptos-labs/aptos-ts-sdk/blob/main/src/api/transaction.ts#L200
```
const transaction = aptos.getSigningMessage({....})
```

**Comment by degenSumi** [comment]
Thanks for the answer @0xmaayan , we already have js sdk so for backwards compatibility as ts sdk does not produce a single bcs signed raw transaction, transaction and signature are separate.
